pkg/cli/admin/release/extract_tools.go

```diff
@@ -908,3 +954,121 @@ func copyAndReplace(errorOutput io.Writer, w io.Writer, r io.Reader, bufferSize
+	}
+
+	config.RequiredFeatureSet = pointer.String(string(data.FeatureSet))
+	config.Profile = pointer.String(manifest.DefaultClusterProfile) // assumption, but there's no install-config data about profile to give us more insight
+	for key := range data.Platform {
+		config.Platform = pointer.String(key)
+	}
+
+	if data.Capabilities != nil {
+		config.Capabilities = &configv1.ClusterVersionCapabilitiesStatus{}
+		if enabled, ok := configv1.ClusterVersionCapabilitySets[data.Capabilities.BaselineCapabilitySet]; !ok {
+			return config, fmt.Errorf("unrecognized baselineCapabilitySet %q", data.Capabilities.BaselineCapabilitySet)
+		} else {
+			if data.Capabilities.BaselineCapabilitySet == configv1.ClusterVersionCapabilitySetCurrent {
+				klog.Infof("If the eventual cluster will not be the same minor version as this %s 'oc', the actual %s capability set may differ.", reportedVersion, data.Capabilities.BaselineCapabilitySet)
+			}
+			config.Capabilities.EnabledCapabilities = append(config.Capabilities.EnabledCapabilities, enabled...)
+		}
+		config.Capabilities.EnabledCapabilities = append(config.Capabilities.EnabledCapabilities, data.Capabilities.AdditionalEnabledCapabilities...)
+
+		klog.Infof("If the eventual cluster will not be the same minor version as this %s 'oc', the known capability sets may differ.", reportedVersion)
+		config.Capabilities.KnownCapabilities = configv1.KnownClusterVersionCapabilities
+	}
+
+	return config, nil
+}
+
+func findClusterIncludeConfig(ctx context.Context, restConfig *rest.Config) (manifestInclusionConfiguration, error) {
+	config := manifestInclusionConfiguration{}
+
+	client, err := configv1client.NewForConfig(restConfig)
+	if err != nil {
+		return config, err
+	}
+
+	if featureGate, err := client.FeatureGates().Get(ctx, "cluster", metav1.GetOptions{}); err != nil {
+		return config, err
+	} else {
+		config.RequiredFeatureSet = pointer.String(string(featureGate.Spec.FeatureSet))
+	}
+
+	if clusterVersion, err := client.ClusterVersions().Get(ctx, "version", metav1.GetOptions{}); err != nil {
+		return config, err
+	} else {
+		config.Overrides = clusterVersion.Spec.Overrides
+		config.Capabilities = &clusterVersion.Status.Capabilities
+
+		// FIXME: eventually pull in GetImplicitlyEnabledCapabilities from https://github.com/openshift/cluster-version-operator/blob/86e24d66119a73f50282b66a8d6f2e3518aa0e15/pkg/payload/payload.go#L237-L240 for cases where a minor update would implicitly enable some additional capabilities.  For now, 4.13 to 4.14 will always enable MachineAPI.
+		currentVersion := clusterVersion.Status.Desired.Version
+		matches := regexp.MustCompile(`^(\d+[.]\d+)[.].*`).FindStringSubmatch(currentVersion)
+		if len(matches) < 2 {
+			return config, fmt.Errorf("failed to parse major.minor version from ClusterVersion status.desired.version %q", currentVersion)
+		} else if matches[1] == "4.13" {
+			machineAPI := configv1.ClusterVersionCapability("MachineAPI")
```

Comment:
This bit (from the earlier FIXME coment through this whole code block) is a ugly hack to work around 4.13-to-4.14 updates implicitly enabling `MachineAPI`.  I hope to drop it soon, and replace it with a [`GetImplicitlyEnabledCapabilities`][1] shifted to library-go.  But we've been through a few rounds of library-go bumps already, and this temporary hack seems reliable enough (to me) for 4.14, with something more robust landing shortly afterwards for 4.15+.

[1]: https://github.com/openshift/cluster-version-operator/blob/86e24d66119a73f50282b66a8d6f2e3518aa0e15/pkg/payload/payload.go#L237-L240

Reply:
Oh, and the reason we can't just ignore `MachineAPI` and assume that 4.13 clusters will already have 4.14-compatible credentials in-cluster is:

```console
$ oc adm release extract --to 4.13 --credentials-requests quay.io/openshift-release-dev/ocp-release:4.13.0-x86_64
$ oc adm release extract --to 4.14 --credentials-requests quay.io/openshift-release-dev/ocp-release:4.14.0-ec.4-x86_64
$ diff -u3 <(yaml2json <4.13/0000_30_machine-api-operator_00_credentials-request.yaml | jq '.[] | select(.metadata.name == "openshift-machine-api-powervs").spec.providerSpec.policies[]') <(yaml2json <4.14/0000_30_machine-api-operator_00_credentials-request.yaml | jq '.[] | select(.metadata.name == "openshift-machine-api-powervs").spec.providerSpec.policies[]')
--- /dev/fd/63  2023-08-09 16:41:04.479985075 -0700
+++ /dev/fd/62  2023-08-09 16:41:04.482985075 -0700
@@ -22,3 +22,16 @@
     "crn:v1:bluemix:public:iam::::role:Viewer"
   ]
 }
+{
+  "attributes": [
+    {
+      "name": "serviceName",
+      "value": "is"
+    }
+  ],
+  "roles": [
+    "crn:v1:bluemix:public:iam::::role:Editor",
+    "crn:v1:bluemix:public:iam::::role:Operator",
+    "crn:v1:bluemix:public:iam::::role:Viewer"
+  ]
+}
```

4.13 PowerVS clusters will need that additional policy before updating to 4.14, and the extraction logic needs to know that after the update to 4.14, `MachineAPI`  will be enabled so it can include that CredentialsRequest.  The only other `MachineAPI` change from 4.13 to 4.14 is `openshift-machine-api-azure` adding `serviceAccountNames`, and I'm not sure if that matters for `ccoctl` or not.

Reply:
Sounds good to me. Can we create a Jira to track the library-go change?

Reply:
I've created [OTA-1010](https://issues.redhat.com/browse/OTA-1010).